MuvrKit/MKWatchConnectivity.swift

```diff
@@ -185,41 +231,80 @@ public final class MKConnectivity : NSObject, WCSessionDelegate {
                     return sample.startDate.timeIntervalSinceDate(lastTime) < 0.04
                 } else {
-                    // first sample: check it is in range
-                    return isInRange(sample)
+                    // the first sample we've seen
+                    return true
                 }
             }
             
+            let documentsUrl = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true).first!
+            let fileUrl = NSURL(fileURLWithPath: documentsUrl).URLByAppendingPathComponent("sensordata.raw")
+            do {
+                try NSFileManager.defaultManager().removeItemAtURL(fileUrl)
+            } catch {
+                //
+            }
+            
+            // we're in a simulator => pretend that we have all data in from..to
             if simulatedSamples {
+                let encoder = MKSensorDataEncoder(target: MKFileSensorDataEncoderTarget(fileUrl: fileUrl), types: recordedTypes, samplesPerSecond: 50)
                 let samples = (0..<sampleCount).map { _ in return Float(0) }
-                return try! MKSensorData(types: recordedTypes, start: from.timeIntervalSince1970, samplesPerSecond: 50, samples: samples)
-            } else {
-                var sampleStart: NSDate? = nil
-                var lastTime: NSDate? = nil
-                return recorder.accelerometerDataFromDate(from, toDate: to).flatMap { (recordedData: CMSensorDataList) -> MKSensorData? in
-                    let samples = recordedData.enumerate().flatMap { (_, e) -> [Float] in
-                        if let data = e as? CMRecordedAccelerometerData where isExpectedSample(data, lastTime: lastTime) {
-                            if sampleStart == nil { // first sample - set range start date
-                                sampleStart = data.startDate
-                            }
-                            lastTime = data.startDate
-                            return [Float(data.acceleration.x), Float(data.acceleration.y), Float(data.acceleration.z)]
-                        }
-                        return []
+                encoder.append(samples)
+                encoder.close(0)
+                NSLog("Generated \(from) - \(to) samples.")
+                return (fileUrl, to)
+            }
+            
+            // try to get the data from the recorder
+            guard let sdl = recorder.accelerometerDataFromDate(from, toDate: to) else {
+                NSLog("No data available.")
+                return nil
+            }
+            var firstSampleTime: NSDate? = nil
+            var lastSampleTime: NSDate? = nil
+            var encoder: MKSensorDataEncoder? = nil
+            // enumerate, creating output only if needed
+            sdl.enumerate().forEach { (_, e) in
+                if let data = e as? CMRecordedAccelerometerData where isInRange(data) {
```

Comment:
We never recover the missing samples: we just skip them, which creates gaps. We might have to interpolate between missing samples instead.